module/validation/seal_validator.go

```diff
@@ -0,0 +1,286 @@
+	sealedID := last.BlockID
+	ancestorID := header.ParentID
+	for ancestorID != sealedID {
+
+		ancestor, err := s.headers.ByBlockID(ancestorID)
+		if err != nil {
+			return nil, fmt.Errorf("could not retrieve ancestor header (%x): %w", ancestorID, err)
+		}
+
+		// keep track of blocks on the fork
+		blockIDs = append(blockIDs, ancestorID)
+
+		payload, err := s.payloads.ByBlockID(ancestorID)
+		if err != nil {
+			return nil, fmt.Errorf("could not get block payload %x: %w", ancestorID, err)
+		}
+
+		// Collect execution results from receipts.
+		for _, receipt := range payload.Receipts {
+			resultID := receipt.ExecutionResult.ID()
+			// incorporatedResults should contain the _first_ appearance of the
+			// ExecutionResult. We are traversing the fork backwards, so we can
+			// overwrite any previously recorded result.
+
+			// ATTENTION:
+			// Here, IncorporatedBlockID (the first argument) should be set
+			// to ancestorID, because that is the block that contains the
+			// ExecutionResult. However, in phase 2 of the sealing roadmap,
+			// we are still using a temporary sealing logic where the
+			// IncorporatedBlockID is expected to be the result's block ID.
+			incorporatedResults[resultID] = flow.NewIncorporatedResult(
+				receipt.ExecutionResult.BlockID,
+				&receipt.ExecutionResult,
+			)
+
+		}
+
+		ancestorID = ancestor.ParentID
+	}
+
+	// Loop forward across the fork and try to create a valid chain of seals.
+	// blockIDs, as populated by the previous loop, is in reverse order.
+	for i := len(blockIDs) - 1; i >= 0; i-- {
+		// if there are no more seals left, we can exit earlier
+		if len(byBlock) == 0 {
+			return last, nil
+		}
+
+		// return an error if there are still seals to consider, but they break
+		// the chain
+		blockID := blockIDs[i]
+		seal, found := byBlock[blockID]
+		if !found {
+			return nil, engine.NewInvalidInputErrorf("chain of seals broken (missing: %x)", blockID)
+		}
+
+		delete(byBlock, blockID)
+
+		// check if we have an incorporatedResult for this seal
+		incorporatedResult, ok := incorporatedResults[seal.ResultID]
+		if !ok {
+			return nil, engine.NewInvalidInputErrorf("seal %x does not correspond to a result on this fork", seal.ID())
+		}
+
+		// check the integrity of the seal
+		err := s.validateSeal(seal, incorporatedResult)
+		if err != nil {
+			if engine.IsInvalidInputError(err) {
+				// Skip fail on an invalid seal. We don't put this earlier in the function
+				// because we still want to test that the above code doesn't panic.
+				// TODO: this is only here temporarily to ease the migration to new chunk
+				// based sealing.
+				if s.requiredChunkApprovals == 0 {
+					log.Warn().Msgf("payload includes invalid seal, continuing validation (%x), %s", seal.ID(), err.Error())
+				} else {
+					return nil, fmt.Errorf("payload includes invalid seal (%x), %w", seal.ID(), err)
+				}
+			} else {
+				return nil, fmt.Errorf("unexpected seal validation error %w", err)
```

Comment:
We usually use a `:` to separate error messages when wrapping:
```suggestion
				return nil, fmt.Errorf("unexpected seal validation error: %w", err)
```